File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/view/ModalView.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.view;

import android.content.Context;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewConfiguration;
import android.widget.FrameLayout;

import com.urbanairship.android.layout.Thomas;
import com.urbanairship.android.layout.environment.Environment;
import com.urbanairship.android.layout.model.BaseModel;
import com.urbanairship.android.layout.model.ModalPresentation;
import com.urbanairship.android.layout.property.ConstrainedSize;
import com.urbanairship.android.layout.property.Margin;
import com.urbanairship.android.layout.property.ModalPlacement;
import com.urbanairship.android.layout.property.Position;
import com.urbanairship.android.layout.util.ConstraintSetBuilder;
import com.urbanairship.android.layout.util.ResourceUtils;
import com.urbanairship.android.layout.widget.ConstrainedFrameLayout;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.constraintlayout.widget.ConstraintSet;
import androidx.core.view.ViewCompat;

public class ModalView extends ConstraintLayout {
    private BaseModel model;
    private ModalPresentation presentation;
    private Environment environment;

    private ConstrainedFrameLayout modalFrame;
    private View containerView;
    private int windowTouchSlop;

    @Nullable private OnClickListener clickOutsideListener = null;

    public ModalView(@NonNull Context context) {
        super(context);
        init(context);
    }

    public ModalView(@NonNull Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public ModalView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    public void init(@NonNull Context context) {
        windowTouchSlop = ViewConfiguration.get(context).getScaledWindowTouchSlop();
    }

    @NonNull
    public static ModalView create(
        @NonNull Context context,
        @NonNull BaseModel model,
        @NonNull ModalPresentation presentation,
        @NonNull Environment environment
    ) {
        ModalView view = new ModalView(context);

        view.setModal(model, presentation, environment);
        return view;
    }

    public void setModal(
        @NonNull BaseModel model,
        @NonNull ModalPresentation presentation,
        @NonNull Environment environment
    ) {
        this.model = model;
        this.presentation = presentation;
        this.environment = environment;

        setId(model.getViewId());
        configureModal();
    }

    public void configureModal() {
        ModalPlacement placement = presentation.getResolvedPlacement(getContext());

        ConstrainedSize size = placement.getSize();
        Position position = placement.getPosition();
        Margin margin = placement.getMargin();
        @ColorInt Integer shadeColor = placement.getShadeColor() != null
            ? placement.getShadeColor().resolve(getContext()) : null;
        makeFrame(size);

        containerView = Thomas.view(getContext(), model, environment);
        FrameLayout.LayoutParams lp = new FrameLayout.LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
        lp.gravity = position != null ? position.getGravity() : Gravity.CENTER;
        if (margin != null) {
            lp.setMargins(margin.getStart(), margin.getTop(), margin.getEnd(), margin.getBottom());
        }
        containerView.setLayoutParams(lp);
        modalFrame.addView(containerView);

        addView(modalFrame);

        int viewId = modalFrame.getId();
        ConstraintSet constraints = ConstraintSetBuilder.newBuilder(getContext())
                                                        .constrainWithinParent(viewId)
                                                        .size(size, viewId)
                                                        .margin(margin, viewId)
                                                        .build();

        if (shadeColor != null) {
            setBackgroundColor(shadeColor);
        }

        constraints.applyTo(this);

        if (environment.isIgnoringSafeAreas()) {
           ViewCompat.setOnApplyWindowInsetsListener(modalFrame, (v, insets) ->
               ViewCompat.dispatchApplyWindowInsets(containerView, insets)
           );
        }
    }

    @Override
    public boolean isOpaque() {
        return false;
    }

    public void makeFrame(ConstrainedSize size) {
        modalFrame = new ConstrainedFrameLayout(getContext(), size);
        modalFrame.setId(generateViewId());
        LayoutParams params = new LayoutParams(LayoutParams.MATCH_CONSTRAINT, LayoutParams.MATCH_CONSTRAINT);
        modalFrame.setLayoutParams(params);
        modalFrame.setElevation(ResourceUtils.dpToPx(getContext(), 16));
    }

    @Override
    public boolean onTouchEvent(@NonNull MotionEvent event) {
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                return true;
            case MotionEvent.ACTION_UP:
                if (isTouchOutside(event) && clickOutsideListener != null) {
                    clickOutsideListener.onClick(this);
                    return true;
                }
                break;
        }

        return super.onTouchEvent(event);
    }

    public void setOnClickOutsideListener(OnClickListener listener) {
        clickOutsideListener = listener;
    }

    private boolean isTouchOutside(@NonNull MotionEvent event) {
        // Get the bounds of the modal
        Rect r = new Rect();
        containerView.getHitRect(r);
        // Expand the bounds by the amount of slop needed to be considered an outside touch
        r.inset(-windowTouchSlop, -windowTouchSlop);

        return !r.contains((int) event.getX(), (int) event.getY());
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/view/ModalView.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.view

import android.content.Context
import android.graphics.Rect
import android.os.Build
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.view.ViewGroup
import android.view.ViewOutlineProvider
import android.widget.FrameLayout
import androidx.annotation.ColorInt
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.constraintlayout.widget.ConstraintLayout.LayoutParams.MATCH_CONSTRAINT
import androidx.constraintlayout.widget.ConstraintLayout.LayoutParams.MATCH_PARENT
import androidx.core.view.ViewCompat
import androidx.core.view.ViewCompat.setOnApplyWindowInsetsListener
import androidx.core.view.WindowInsetsCompat
import com.urbanairship.android.layout.ModalPresentation
import com.urbanairship.android.layout.environment.ViewEnvironment
import com.urbanairship.android.layout.model.AnyModel
import com.urbanairship.android.layout.util.ConstraintSetBuilder
import com.urbanairship.android.layout.util.LayoutUtils
import com.urbanairship.android.layout.util.ResourceUtils
import com.urbanairship.android.layout.widget.ClippableFrameLayout
import com.urbanairship.android.layout.widget.ConstrainedFrameLayout

internal class ModalView(
    context: Context,
    model: AnyModel,
    presentation: ModalPresentation,
    viewEnvironment: ViewEnvironment
) : ConstraintLayout(context) {

    private val windowTouchSlop by lazy {
        ViewConfiguration.get(context).scaledWindowTouchSlop
    }

    private val modalFrame: ViewGroup

    /** The modal frame's min height as declared by the layout (e.g. `min_height: 100%`). */
    private val initialModalFrameMinHeight: Int

    private var lastAppliedImeBottom: Int = 0

    private var clickOutsideListener: OnClickListener? = null

    init {
        val placement = presentation.getResolvedPlacement(context)
        val size = placement.size
        val position = placement.position
        val margin = placement.margin
        @ColorInt val shadeColor = placement.shadeColor?.resolve(context)

        modalFrame = ConstrainedFrameLayout(context, size).apply {
            id = generateViewId()
            layoutParams = LayoutParams(MATCH_CONSTRAINT, MATCH_CONSTRAINT).apply {
                margin?.let {
                    setMargins(
                        ResourceUtils.dpToPx(context, it.start).toInt(),
                        ResourceUtils.dpToPx(context, it.top).toInt(),
                        ResourceUtils.dpToPx(context, it.end).toInt(),
                        ResourceUtils.dpToPx(context, it.bottom).toInt()
                    )
                }
            }
            clipChildren = false
            clipToPadding = false
            outlineProvider = ViewOutlineProvider.BOUNDS

            LayoutUtils.applyBorderAndBackground(this,null, placement.border, placement.backgroundColor)

            placement.shadow?.androidShadow?.let { shadow ->
                applyShadow(this, shadow.color.resolve(context), shadow.elevation)
            }
        }

        val containerView = ClippableFrameLayout(context).apply {
            layoutParams = FrameLayout.LayoutParams(MATCH_PARENT, MATCH_PARENT)

            placement.border
                ?.innerRadii { ResourceUtils.dpToPx(context, it) }
                ?.let {
                    setClipPathBorderRadius(it)
                }
        }

        val contentView = model.createView(context, viewEnvironment, null)
        containerView.addView(contentView)

        modalFrame.addView(containerView)
        addView(modalFrame)

        val viewId = modalFrame.id
        val constraints = ConstraintSetBuilder.newBuilder(context)
            .constrainWithinParent(viewId)
            .size(size, placement.shouldIgnoreSafeArea(), viewId)
            .position(position, viewId)
            .build()

        constraints.applyTo(this)
        initialModalFrameMinHeight =
            constraints.getConstraint(modalFrame.id).layout.heightMin.coerceAtLeast(0)
        shadeColor?.let { setBackgroundColor(it) }

        if (viewEnvironment.isIgnoringSafeAreas) {
            setOnApplyWindowInsetsListener(modalFrame) { _: View, insets: WindowInsetsCompat ->
                ViewCompat.dispatchApplyWindowInsets(containerView, insets)
            }
        }
    }

    override fun isOpaque(): Boolean = false

    override fun onTouchEvent(event: MotionEvent): Boolean {
        when (event.action) {
            MotionEvent.ACTION_DOWN ->
                return true
            MotionEvent.ACTION_UP ->
                if (isTouchOutside(event) && clickOutsideListener != null) {
                    clickOutsideListener?.onClick(this)
                    return true
                }
        }
        return super.onTouchEvent(event)
    }

    fun setOnClickOutsideListener(listener: OnClickListener?) {
        clickOutsideListener = listener
    }

    /**
     * Called by [ModalActivity] when the IME bottom inset changes.
     *
     * The activity already accounts for the IME by adding it to the ModalView's bottom
     * padding, so we don't need to move the modal frame ourselves here. What we *do*
     * need to do is prevent the frame's declared `min_height` (e.g. `100%`) from
     * forcing the MATCH_CONSTRAINT height to overflow the shrunken inner area. We
     * clamp `matchConstraintMinHeight` directly on the LayoutParams so the change
     * lands in the current measure pass without relying on a ConstraintSet
     * clone/applyTo round-trip (which wasn't reliably round-tripping the override).
     */
    internal fun applyKeyboardInset(imeBottom: Int) {
        if (lastAppliedImeBottom == imeBottom) return
        lastAppliedImeBottom = imeBottom

        clampModalFrameMinHeight()
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        // On API < 28 without ignoreSafeArea, adjustResize shrinks this view when the IME
        // shows, but onApplyWindowInsets is not re-dispatched. Re-clamp here so the modal
        // frame's min height tracks the resized window.
        post {
            clampModalFrameMinHeight()
        }
    }

    private fun clampModalFrameMinHeight() {
        val lp = modalFrame.layoutParams as? LayoutParams ?: return
        if (height <= 0) return

        // The space available to the modal frame inside the ModalView (after padding
        // and the frame's own margins). When the IME is up, ModalView.paddingBottom
        // already includes the IME height, so this naturally shrinks.
        val availableHeight =
            height - paddingTop - paddingBottom - lp.topMargin - lp.bottomMargin

        val newMinHeight = when {
            availableHeight <= 0 -> initialModalFrameMinHeight
            availableHeight < initialModalFrameMinHeight -> availableHeight
            else -> initialModalFrameMinHeight
        }

        if (lp.matchConstraintMinHeight != newMinHeight) {
            lp.matchConstraintMinHeight = newMinHeight
            modalFrame.layoutParams = lp
        }
    }

    private fun applyShadow(view: View, color: Int, elevation: Float) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            view.outlineAmbientShadowColor = color
            view.outlineSpotShadowColor = color
        }
        view.elevation = ResourceUtils.dpToPx(context, elevation.toInt())
    }

    private fun isTouchOutside(event: MotionEvent): Boolean {
        // Get the bounds of the modal
        val r = Rect()
        modalFrame.getHitRect(r)
        // Expand the bounds by the amount of slop needed to be considered an outside touch
        r.inset(-windowTouchSlop, -windowTouchSlop)
        return !r.contains(event.x.toInt(), event.y.toInt())
    }
}
